docs/docs/typelevel.md

```diff
@@ -0,0 +1,532 @@
+easy to see that we use the same recursive decomposition of `n` in
+both cases and that the result types of each case correspond. But to
+prove this correspondence in general would require some sort of
+symbolic evaluation and it is not yet clear what the details of that
+would be.
+
+But there is a simpler way: We can _delay_ checking the precise result
+type of the body of an `macro` until the macro is inlined.  The
+declared result type can be fully unfolded based on the static
+information available at the inlining point. At that point the inlined
+expansion of the macro is type-checked using the fully computed types.
+
+That still begs the question how we are going to typecheck the
+_definition_ of a macro. The idea is to approximate any computed
+type at the definition that depends on unknown parameters by its upper
+bound.
+
+E.g., in the case of `toNat`, the type of the recursive call
+`toNat(n - 1)` would be the upper bound `Nat` of the declared return type
+`ToNat(n - 1)`. This gives `S[Nat]` as the type of the else
+clause. The full type of the right hand side `Z | S[Nat]` would then
+be checked against the upper approximation of the declared result type
+`ToNat(n)`, which is again `Nat`. This succeeds, so the definition is deemed type-correct.
+
+
+### Second Step: Nicer Syntax for Type Definitions
+
+The syntax for `ToNat` was a bit clunky. This is no outlier - it turns out
+that most inline type definitions use similar staging constructs. All
+parameters tend to be inline, and the body is typically a conditional
+expression which returns a quoted type in each branch. We can cut down
+this boilerplate by introducing a shorthand form for parameterized
+type definitions, which allows us to express `ToNat` as follows:
+
+    type ToNat(n: Int) <: Nat =
+      if n == 0 then Z
+      else S[ToNat(n - 1)]
+
+The short form expands precisely to the previous definition of `ToNat`.
+To get from a short form type definition to a long form type macro,
+the following rewritings are performed:
+
+ - `type` is expanded to `inline type def`.
+ - All value parameters get an `inline` modifier.
+ - If the right hand side consists of a conditional, it is put in a spliced block `~{...}`.
+ - The right hand sides of all conditionals are put in type quotes `'[...]`.
+
+We expect that most computed types coming up in practice will use the
+short form. The long form is kept around, to cater for more advanced
+use cases, and to give a precise meaning to the short form.
+
+### Third Step: Query Conditionals
+
+What about defining the inverse of `ToNat`? This should be a function
+that takes a _type_ of peano numbers and produces a compile-time
+integer _value_. We can emulate the effect of such a function with
+implicits but it's a bit roundabout since we would need several
+overloaded methods each taking a different implicit parameter. A more
+direct, closed formulation is possible if we allow implicit searches
+as conditions in spliced code. Example:
```

Comment:
If the goal is to avoid needing logic programming, why not allow matching on types? Allowing for implicit searches is pretty interesting idea, but detracts from that goal... and requiring implicit search to use pattern-matching/unification on types seems pretty roundabout?

Reply:
Matching on types is problematic because that's super fragile for a rich type system with subtyping. This observation was the blocker why we did not do any of this before. Implicit search overcomes this blocker, so is IMO really important.